File: Day1-9/8.py
import operator
from collections import defaultdict

instructions = {
    'inc': operator.add,
    'dec': operator.sub
}

comparators = {
    '>': operator.gt,
    '<': operator.lt,
    '>=': operator.ge,
    '<=': operator.le,
    '==': operator.eq,
    '!=': operator.ne
}


def run_instructions(instructions):
    registers = defaultdict(int)
    max_at_runtime = 0
    for i in instructions:
        asn_reg, op, amt, cmp_reg, cmp_op, cmp_amt = i
        if cmp_op(registers[cmp_reg], cmp_amt):
            registers[asn_reg] = op(registers[asn_reg], amt)
        max_at_runtime = max(max_at_runtime, registers[asn_reg])
    max_at_end = max(registers.values())
    return (max_at_end, max_at_runtime)


def parse_instruction(line):
    tokens = line.split()
    return (
        tokens[0],
        instructions[tokens[1]],
        int(tokens[2]),
        tokens[4],
        comparators[tokens[5]],
        int(tokens[6])
    )


if __name__ == '__main__':
    with open('8.txt') as f:
        lines = f.readlines()
    instructions = [parse_instruction(l) for l in lines]
    max_at_end, max_at_runtime = run_instructions(instructions)
    print("Part 1: {}".format(max_at_end))
    print("Part 2: {}".format(max_at_runtime))

File: Day1-9/8.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

type Instruction struct {
	assignmentReg   string
	assignmentOp    string
	assignmentAmt   int
	compareReg      string
	compareOperator string
	compareAmt      int
}

func assignmentOp(operator string, a int, b int) int {
	switch operator {
	case "inc":
		return a + b
	case "dec":
		return a - b
	}
	panic("Invalid assignment operator")
}

func compareOp(operator string, a int, b int) bool {
	switch operator {
	case ">":
		return a > b
	case "<":
		return a < b
	case "<=":
		return a <= b
	case ">=":
		return a >= b
	case "==":
		return a == b
	case "!=":
		return a != b
	}
	panic("Invalid assignment operator")
}

func parseLine(line string) Instruction {
	fields := strings.Fields(line)
	inst := Instruction{}
	inst.assignmentReg = fields[0]
	inst.assignmentOp = fields[1]
	inst.assignmentAmt, _ = strconv.Atoi(fields[2])
	inst.compareReg = fields[4]
	inst.compareOperator = fields[5]
	inst.compareAmt, _ = strconv.Atoi(fields[6])
	return inst
}

func runInterpreter(instructions []Instruction) (maxAtEnd int, maxAtRuntime int) {
	registers := make(map[string]int)
	maxAtRuntime = 0
	maxAtEnd = 0
	for _, inst := range instructions {
		// Check that assignment register has been initialized
		if _, ok := registers[inst.assignmentReg]; !ok {
			registers[inst.assignmentReg] = 0
		}
		// Check that comparing register has been initialized
		if _, ok := registers[inst.compareReg]; !ok {
			registers[inst.compareReg] = 0
		}

		compareRegVal := registers[inst.compareReg]
		if compareOp(inst.compareOperator, compareRegVal, inst.compareAmt) {
			assignmentRegVal := registers[inst.assignmentReg]
			assignmentVal := assignmentOp(inst.assignmentOp, assignmentRegVal, inst.assignmentAmt)
			registers[inst.assignmentReg] = assignmentVal

			// Update max assignment value during runtime
			if assignmentVal > maxAtRuntime {
				maxAtRuntime = assignmentVal
			}
		}
	}

	for _, val := range registers {
		if val > maxAtEnd {
			maxAtEnd = val
		}
	}

	return
}

func main() {
	dat, _ := ioutil.ReadFile("./8.txt")
	lines := strings.Split(string(dat), "\n")
	instructions := make([]Instruction, len(lines))

	for i, line := range lines {
		instructions[i] = parseLine(line)
	}

	maxAtEnd, maxAtRuntime := runInterpreter(instructions)
	fmt.Printf("Part 1: %d\n", maxAtEnd)
	fmt.Printf("Part 2: %d\n", maxAtRuntime)
}
